azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -0,0 +1,190 @@
+#
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.version import get_distro
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.event import WALAEventOperation, add_event
+from azurelinuxagent.common import conf
+import platform
+
+# Define support matrix for Regorus and policy engine feature.
+# Dict in the format: { distro:min_supported_version }
+POLICY_SUPPORTED_DISTROS = {
+    'ubuntu': FlexibleVersion('16.04'),
+    'mariner': FlexibleVersion('2')
+}
+POLICY_SUPPORTED_ARCHITECTURE = ['x86_64', 'amd64']
+
+
+# Common logging function across PolicyEngineConfigurator and PolicyEngine classes,
+# so should be a module-level function.
+def log_policy(formatted_string, is_success=True, op=WALAEventOperation.Policy, send_event=True):
+    """
+    Log information to console and telemetry.
+    """
+    if is_success:
+        logger.info("[Policy] " + formatted_string)
+    else:
+        logger.error("[Policy] " + formatted_string)
+    if send_event:
+        add_event(op=op, message=formatted_string, is_success=is_success)
+
+
+class PolicyEngineConfigurator:
+    """
+    This Singleton class is responsible for checking if policy is enabled for the agent.
+    Policy can only be enabled/disabled via configuration file, and can only be enabled on supported distros.
+    If an error occurs during Regorus import, policy will be globally disabled for the agent via the _policy_enabled flag.
+    Use: PolicyEngineConfigurator.get_instance()
+    """
+    _instance = None  # configurator is implemented as a singleton
+    _initialized = False  # set to true on first init, even if Regorus import fails, we don't want to retry.
+    _policy_enabled = False
+
+    def __init__(self):
+        # we only attempt to import Regorus if it is both enabled via conf and supported by the platform
+        try:
+            if PolicyEngineConfigurator._initialized:
+                return
+
+            if not self.is_policy_configured():
+                log_policy("Policy enforcement is disabled.")
+                return
+
+            if not self._is_policy_supported():
+                # If distro is unsupported but feature is enabled, we never set _policy_enabled=True.
+                # Any methods will be no-ops, and query evaluation will return an empty allowlist
+                # so no extensions will be allowed.
+                log_policy("Policy enforcement is unsupported on this platform.")
+                return
+
+            # Regorus import should only be attempted after completing the above checks within
+            # the configurator, but the module itself needs to be accessible outside this class.
+            global regorus  # pylint: disable=global-statement
+            import azurelinuxagent.ga.policy.regorus as regorus
+            PolicyEngineConfigurator._policy_enabled = True
+
+        except (ImportError, NameError) as ex:
```

Comment:
How could ImportError and NameError happen?

Reply:
Removed these errors, not relevant since we're no longer using a python wrapper